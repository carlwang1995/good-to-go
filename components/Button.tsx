"use client";

type ButtonProps = {
  title: string | React.JSX.Element;
  type: "confirm" | "cancel" | "close" | "undone";
  onSmash?: () => void;
};

const Button = ({ title, type, onSmash }: ButtonProps) => {
  let style;
  if (type === "confirm") {
    style = "btn_blue";
  } else if (type === "cancel") {
    style = "btn_white";
  } else if (type === "close") {
    style = "btn_red";
  } else if (type === "undone") {
    style = "btn_undone";
  }
  return (
    <button className={style} onClick={onSmash}>
      {title}
    </button>
  );
};

export default Button;
