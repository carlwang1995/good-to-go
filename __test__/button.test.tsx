import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Button from "@/components/Button";

describe("測試Button是否正常", () => {
  test("測試渲染結果與 Snapshot 一致", () => {
    let { container } = render(<Button title="測試" type="confirm" />);
    expect(container).toMatchSnapshot();
  });
  test("當按下按鈕時，確定 mockFunc 會被呼叫", async () => {
    const mockFunc = jest.fn(); // 宣告一個模擬用的函式

    render(
      // 將 Button 渲染出來，因為之後才可以被抓得到
      <Button title="測試" type="confirm" onSmash={mockFunc} />,
    );

    const testButton = screen.getByText("測試");
    await userEvent.click(testButton); // 模擬使用者的點擊行為
    expect(mockFunc).toBeCalledTimes(1);
  });
});
