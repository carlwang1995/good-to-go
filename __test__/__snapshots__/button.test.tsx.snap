// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`測試Button是否正常 測試渲染結果與 Snapshot 一致 1`] = `
<div>
  <button
    class="btn_blue"
  >
    測試
  </button>
</div>
`;
